experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,131 @@
+            request_sender,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (sender, receiver) = oneshot::channel();
+
+        let request = Message {
+            data: request,
+            response_sender: sender,
+        };
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut current_requests = self
+                .current_requests
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+            current_requests.push(request);
+
+            if current_requests.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(current_requests.deref_mut(), vec![]);
+
+                // Shuffle current requests and send them to the backend.
+                let request_sender = self.request_sender.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_sender).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        receiver.await.context("Couldn't receive response")
+    }
+
+    // Shuffles requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(mut requests: Vec<Message>, request_sender: F) -> anyhow::Result<()> {
+        requests.shuffle(&mut thread_rng());
+
+        let response_futures: Vec<_> = requests
+            .drain(..)
+            .into_iter()
+            .map(|request| {
+                let request_sender_clone = request_sender.clone();
+                async move {
+                    let response = (request_sender_clone)(request.data).await;
+                    Message {
+                        data: response,
+                        response_sender: request.response_sender,
+                    }
+                }
+            })
+            .collect();
+        let responses = join_all(response_futures).await;
+
+        Self::shuffle_responses(responses)
+    }
+
+    // Shuffles received backend responses and sends them back to the client async tasks using
+    // corresponding channels defined in the [`Message::response_sender`].
+    fn shuffle_responses(mut responses: Vec<Message>) -> anyhow::Result<()> {
+        responses.shuffle(&mut thread_rng());
+
+        for response in responses.drain(..) {
+            response
+                .response_sender
+                .send(response.data.clone())
+                .map_err(|error| anyhow!("Couldn't send the response: {:?}", error))?;
+        }
+        Ok(())
```

Comment:
I think this for loop could be something like this expression (which would be the return value of the method itself)

```rust
responses
  .into_iter()
  .map(|Message { data, response_sender }| response_sender.send(data))
  .collect()
  .map(|_: Vec<()>| ())
```

Reply:
It will have to look like this:
```rust
responses
    .into_iter()
    .map(|Message { data, response_sender }| response_sender.send(data))
    .collect::<Vec<anyhow::Result<(), Vec<u8>>>>()
    .map(|_: Vec<anyhow::Result<(), Vec<u8>>>| ())
```
And even then it still doesn't want to build the vec:
```
error[E0599]: the method `map` exists for struct `Vec<Result<(), Vec<u8>>>`, but its trait bounds were not satisfied
   --> experimental/trusted_shuffler/trusted_shuffler/src/lib.rs:143:14
    |
143 |             .map(|_: Vec<anyhow::Result<(), Vec<u8>>>| ())
    |              ^^^ method cannot be called on `Vec<Result<(), Vec<u8>>>` due to unsatisfied trait bounds
    |
   ::: /usr/local/cargo/toolchains/nightly-2022-02-17-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/vec/mod.rs:400:1
    |
400 | pub struct Vec<T, #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global> {
    | ------------------------------------------------------------------------------------------------ doesn't satisfy `Vec<Result<(), Vec<u8>>>: Iterator`
```